File: textdb/textdb-api/src/main/java/edu/uci/ics/textdb/api/schema/Schema.java
package edu.uci.ics.textdb.api.schema;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import edu.uci.ics.textdb.api.constants.JsonConstants;
import edu.uci.ics.textdb.api.exception.TextDBException;

public class Schema {
    
    private final List<Attribute> attributes;
    private final Map<String, Integer> nameLowercaseIndexMap;

    public Schema(Attribute... attributes) {
        this(Arrays.asList(attributes));
    }
    
    @JsonCreator
    public Schema(
            @JsonProperty(value = JsonConstants.ATTRIBUTES, required = true)
            List<Attribute> attributes) {
        if (attributes.isEmpty()) {
            throw new TextDBException("Cannot create schema: attributes is empty");
        }
        this.attributes = new ArrayList<>(attributes);
        this.nameLowercaseIndexMap = new HashMap<>();
        
        for (int i = 0; i < attributes.size(); i++) {
            String attributeNameLowercase = attributes.get(i).getAttributeName().toLowerCase();
            if (nameLowercaseIndexMap.containsKey(attributeNameLowercase)) {
                throw new TextDBException("Cannot create schema: duplicate attribute name " + attributeNameLowercase + 
                        ". (attribute name is case insensitive)");
            }
            nameLowercaseIndexMap.put(attributeNameLowercase, i);
        }
    }

    @JsonProperty(value = JsonConstants.ATTRIBUTES)
    public List<Attribute> getAttributes() {
        return new ArrayList<>(attributes);
    }
    
    @JsonIgnore
    public int size() {
        return attributes.size();
    }
    
    @JsonIgnore
    public List<String> getAttributeNames() {
        return attributes.stream().map(attr -> attr.getAttributeName()).collect(Collectors.toList());
    }

    public Integer getIndex(String attributeName) {
        String attributeNameLowercase = attributeName.trim().toLowerCase();
        if (! nameLowercaseIndexMap.containsKey(attributeNameLowercase)) {
            throw new TextDBException(String.format(
                    "Cannot find the attribute: %s does not exist in the schema %s", 
                    attributeName, this.getAttributeNames()));
        }
        return nameLowercaseIndexMap.get(attributeNameLowercase);
    }
    
    public Attribute getAttribute(String attributeName) {
        return attributes.get(getIndex(attributeName));
    }
    
    public Attribute getAttribute(int index) {
        if (index < attributes.size()) {
            return attributes.get(index);
        }
        throw new TextDBException(String.format(
                "Cannot get attribute at index %d: schema only has %d attributes", 
                index, attributes.size()));
    }

    @JsonIgnore
    public boolean containsField(String attributeName) {
        String attributeNameLowercase = attributeName.trim().toLowerCase();
        return nameLowercaseIndexMap.containsKey(attributeNameLowercase);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((attributes == null) ? 0 : attributes.hashCode());
        result = prime * result + ((nameLowercaseIndexMap == null) ? 0 : nameLowercaseIndexMap.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object toCompare) {
        if (this == toCompare)
            return true;
        if (toCompare == null)
            return false;
        if (getClass() != toCompare.getClass())
            return false;
        
        Schema that = (Schema) toCompare;  
        return this.attributes.equals(that.attributes) 
                && this.nameLowercaseIndexMap.equals(that.nameLowercaseIndexMap);
    }
}

File: core/workflow-core/src/main/scala/edu/uci/ics/amber/core/tuple/Schema.scala
package edu.uci.ics.amber.core.tuple

import com.fasterxml.jackson.annotation.{JsonCreator, JsonIgnore, JsonProperty}
import com.google.common.base.Preconditions.checkNotNull

import scala.collection.immutable.ListMap

case class Schema @JsonCreator() (
                                   @JsonProperty(value = "attributes", required = true) attributes: List[Attribute] = List()
                                 ) extends Serializable {

  checkNotNull(attributes)

  private val attributeIndex: Map[String, Int] =
    attributes.view.map(_.getName.toLowerCase).zipWithIndex.toMap

  def this(attrs: Attribute*) = this(attrs.toList)

  @JsonProperty(value = "attributes")
  def getAttributes: List[Attribute] = attributes

  @JsonIgnore
  def getAttributeNames: List[String] = attributes.map(_.getName)

  def getIndex(attributeName: String): Int = {
    if (!containsAttribute(attributeName)) {
      throw new RuntimeException(s"$attributeName is not contained in the schema")
    }
    attributeIndex(attributeName.toLowerCase)
  }

  def getAttribute(attributeName: String): Attribute = attributes(getIndex(attributeName))

  @JsonIgnore
  def containsAttribute(attributeName: String): Boolean =
    attributeIndex.contains(attributeName.toLowerCase)

  override def hashCode(): Int = attributes.hashCode

  override def equals(obj: Any): Boolean =
    obj match {
      case that: Schema => this.attributes == that.attributes
      case _            => false
    }

  override def toString: String = s"Schema[$attributes]"

  def getPartialSchema(attributeNames: List[String]): Schema = {
    Schema(attributeNames.map(name => getAttribute(name)))
  }

  /**
    * This method converts the schema into a raw format where each attribute name
    * and attribute type are represented as strings. This is for serialization between languages.
    */
  def toRawSchema: Map[String, String] =
    attributes.foldLeft(ListMap[String, String]())((list, attr) =>
      list + (attr.getName -> attr.getType.name())
    )

  /**
    * Creates a new Schema by adding attributes to the current Schema.
    */
  def add(attributesToAdd: Iterable[Attribute]): Schema = {
    val newAttributes = attributes ++ attributesToAdd
    Schema(newAttributes)
  }

  /**
    * Creates a new Schema by removing attributes from the current Schema.
    */
  def remove(attributeNames: Iterable[String]): Schema = {
    val attributesToRemove = attributeNames.map(_.toLowerCase).toSet
    val remainingAttributes = attributes.filterNot(attr =>
      attributesToRemove.contains(attr.getName.toLowerCase)
    )
    Schema(remainingAttributes)
  }

  /**
    * Creates a new Schema by adding a single attribute.
    */
  def add(attribute: Attribute): Schema = add(List(attribute))

  /**
    * Creates a new Schema by adding a single attribute.
    */
  def add(attributeName: String, attributeType: AttributeType): Schema = add(List(new Attribute(attributeName, attributeType)))

  /**
    * Creates a new Schema by adding a Schema.
    */
  def add(schema: Schema): Schema = {
    add(schema.attributes)
  }

  /**
    * Creates a new Schema by removing a single attribute.
    */
  def remove(attributeName: String): Schema = remove(List(attributeName))
}

object Schema {

  /**
    * Creates a Schema instance from a raw schema map.
    */
  def fromRawSchema(raw: Map[String, String]): Schema = {
    Schema(raw.map { case (name, attrType) =>
      new Attribute(name, AttributeType.valueOf(attrType))
    }.toList)
  }
}